framework/src/org/checkerframework/common/value/ValueCheckerUtils.java

```diff
@@ -149,31 +151,38 @@ public static Range getRangeFromValues(List<? extends Number> values) {
             return values;
         }
-        if (expectedType == Integer.class
-                || expectedType == int.class
-                || expectedType == Long.class
-                || expectedType == long.class
-                || expectedType == Short.class
-                || expectedType == short.class
-                || expectedType == Byte.class
-                || expectedType == byte.class) {
-            for (Long value = range.from; value <= range.to; value++) {
-                values.add(expectedType.cast(value.longValue()));
-            }
-        } else if (expectedType == Double.class
-                || expectedType == double.class
-                || expectedType == Float.class
-                || expectedType == float.class) {
-            for (Long value = range.from; value <= range.to; value++) {
-                values.add(expectedType.cast(value.doubleValue()));
-            }
-        } else if (expectedType == Character.class || expectedType == char.class) {
-            for (Long value = range.from; value <= range.to; value++) {
-                values.add(expectedType.cast((char) value.intValue()));
+
+        // The subtraction does not overflow, because the width has already been checked, so the
+        // bound difference is less than ValueAnnotatedTypeFactory.MAX_VALUES.
+        long boundDifference = range.to - range.from;
+
+        // Each value is computed as a sum of the first value and an offset within the range,
+        // to avoid having range.to as an upper bound of the loop. range.to can be Long.MAX_VALUE,
+        // in which case a comparison value <= range.to would be always true.
+        // boundDifference is always much smaller than Long.MAX_VALUE
+        for (long offset = 0; offset <= boundDifference; offset++) {
+            long value = range.from + offset;
+
+            Object convertedValue;
+            if (expectedType == Integer.class) {
+                convertedValue = (int) value;
+            } else if (expectedType == Short.class) {
+                convertedValue = (short) value;
+            } else if (expectedType == Byte.class) {
+                convertedValue = (byte) value;
+            } else if (expectedType == Long.class) {
+                convertedValue = value;
+            } else if (expectedType == Double.class) {
+                convertedValue = (double) value;
+            } else if (expectedType == Float.class) {
+                convertedValue = (float) value;
+            } else if (expectedType == Character.class) {
+                convertedValue = (char) value;
+            } else {
+                throw new UnsupportedOperationException(
+                        "ValueCheckerUtils: unexpected class: " + expectedType);
             }
-        } else {
-            throw new UnsupportedOperationException(
-                    "ValueCheckerUtils: unexpected class: " + expectedType);
+            values.add(expectedType.cast(convertedValue));
```

Comment:
`Class.cast` throws `ClassCastException` if the argument is not of the correct type, so a cast is needed to perform the conversion. `Class.cast` is used for casting to `T` without needing to suppress an unchecked cast warning.
I thought it would be too much duplication to have the loop 7 times there. If this code is not readable enough, I think extracting conversion logic to a separate method should help.